src/fromager/sdist.py

```diff
@@ -93,16 +101,17 @@ def handle_requirement(
     # Update the dependency graph after we determine that this requirement is
     # useful but before we determine if it is redundant so that we capture all
     # edges to use for building a valid constraints file.
-    ctx.dependency_graph.add_dependency(
-        parent_name=canonicalize_name(why[-1][1].name) if why else None,
-        parent_version=why[-1][2] if why else None,
-        req_type=req_type,
-        req=req,
-        req_version=Version(str(resolved_version)),
-        download_url=source_url,
-        pre_built=pre_built,
-    )
-    ctx.write_to_graph_to_file()
+    if req_type != RequirementType.TOP_LEVEL:
```

Comment:
Maybe `add_dependency()` should be idempotent instead of asking the caller to check?

Reply:
I could add checks to make sure there are no duplicate edges, but in our current downstream we allow duplicate edges to show up

Reply:
If we only need to prevent the top level duplicates, it's OK to just do the check here. Maybe expand the comment to mention that we're expecting top level edges to already exist because they were all resolved before starting to bootstrap, like you have on line 289.